src/VirtoCommerce.CatalogModule.Data/Services/CatalogService.cs

```diff
@@ -184,6 +189,43 @@ private void ClearCache(Catalog[] catalogs)
             {
                 CatalogTreeCacheRegion.ExpireTokenForKey(catalog.Id);
             }
-        }
+        }
+
+        private CatalogChangedEntriesAggregate GetDeletedEntries(Catalog[] catalogs)
+        {
+            using (var repository = _repositoryFactory())
+            {
+                var catalogChangedEntries = catalogs.Select(x => new GenericChangedEntry<Catalog>(x, EntryState.Deleted)).ToList();
+
+                var foundCatalogIds = catalogs.Select(x => x.Id).ToList();
+                var categoryIds = repository.Categories.Where(x => foundCatalogIds.Contains(x.CatalogId)).Select(x => x.Id).ToList();
+                var productIds = repository.Items.Where(x => foundCatalogIds.Contains(x.CatalogId)).Select(x => x.Id).ToList();
+
+                var categoryChangedEntries = categoryIds.Select(id =>
+                {
+                    var category = AbstractTypeFactory<Category>.TryCreateInstance();
+                    category.Id = id;
+                    var entry = new GenericChangedEntry<Category>(category, EntryState.Deleted);
+
+                    return entry;
+                }).ToList();
+
+                var productChangedEntries = productIds.Select(id =>
+                {
+                    var product = AbstractTypeFactory<CatalogProduct>.TryCreateInstance();
+                    product.Id = id;
+                    var entry = new GenericChangedEntry<CatalogProduct>(product, EntryState.Deleted);
+
+                    return entry;
+                }).ToList();
+
+                return new CatalogChangedEntriesAggregate
+                {
+                    CatalogEntries = catalogChangedEntries,
+                    CategoryEntries = categoryChangedEntries,
+                    ProductEntries = productChangedEntries,

```

Comment:
If we use `CatalogChangedEntriesAggregate` only in the private method and we don`t need to make this aggregate extendable:

1. Change collection type to List<T> instead of IEnumerable<T> (increase performance by removing casts operations);
2. Make `CatalogChangedEntriesAggregate` sealed.